rnd/autogpt_server/autogpt_server/data/block.py

```diff
@@ -0,0 +1,130 @@
+import inspect
+import json
+
+from abc import ABC, abstractmethod
+from prisma.models import AgentBlock
+from pydantic import BaseModel
+from typing import ClassVar
+
+
+class Block(ABC, BaseModel):
+    @property
+    @abstractmethod
+    def input_schema(self) -> dict[str, str]:
+        """
+        The schema for the block input data. The keys are the names of the input,
+        and the values are the types of the input.
+        Example:
+        {
+            "system_prompt": "str",
+            "user_prompt": "str",
+            "max_tokens": "int",
+        }
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def output_schema(self) -> dict[str, str]:
+        """
+        The schema for the block possible output. The keys are the names of the output,
+        and the values are the types of the output.
+        Example:
+        {
+            "on_completion": "str",
+            "on_failure": "str",
+        }
+        """
+        pass
+
+    @abstractmethod
+    def run(self, input_data: dict[str, str]) -> (str, str):
+        """
+        Run the block with the given input data.
+        Args:
+            input_data: The input data with the structure of input_schema.
+        Returns:
+            The output data, with the structure of one entry of the output_schema.
+        """
+        pass
+
+    async def execute(self, input_data: dict[str, str]) -> dict[str, str]:
+        result = self.run(input_data)
+        if inspect.isawaitable(result):
+            return await result
+        return result
+
+
+# ===================== Inline-Block Implementations ===================== #
+
+
+class ParrotBlock(Block):
```

Comment:
These are a couple of sample implementations (used for testing).
next is integrating constrained AutoGPT agent into this.